ttnn/cpp/ttnn/operations/experimental/dropout/device/dropout_device_operation.cpp

```diff
@@ -0,0 +1,152 @@
+            static_cast<int>(input_tensor.get_layout()));
+
+        TT_FATAL(
+            input_tensor.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED,
+            "Dropout operation requires Interleaved memory layout when working with non-sharded input tensor. Input "
+            "memory layout: `{}`",
+            static_cast<int>(input_tensor.memory_config().memory_layout));
+    }
+
+    if (preallocated_output_tensor.has_value()) {
+        const auto computed_output_shape = compute_output_specs(args, tensor_args).logical_shape();
+        const auto preallocated_output_shape = preallocated_output_tensor.value().get_logical_shape();
+        TT_FATAL(
+            preallocated_output_shape == computed_output_shape,
+            "When preallocted output tensor is used, Dropout operation requires its shape to match the computed "
+            "shape. Computed shape: {}, Shape in preallocated output tensor: {}",
+            computed_output_shape,
+            preallocated_output_shape);
+
+        if (!input_tensor.is_sharded()) {
+            TT_FATAL(
+                (preallocated_output_tensor.value().get_layout() == Layout::TILE),
+                "Dropout operation requires output tensor to be in Tile layout when working with non-sharded tensor.");
+        }
+    }
+}
+
+spec_return_value_t DropoutDeviceOperation::compute_output_specs(
+    const operation_attributes_t& args, const tensor_args_t& tensor_args) {
+    if (tensor_args.preallocated_output.has_value()) {
+        return tensor_args.preallocated_output->get_tensor_spec();
+    }
+
+    auto output_layout = Layout::TILE;
+    if (args.output_memory_config.is_sharded()) {
+        output_layout = tensor_args.input.get_layout();
+    }
+
+    const auto output_shape = tensor_args.input.logical_shape();
+    return TensorSpec(output_shape, TensorLayout(args.output_dtype, output_layout, args.output_memory_config));
+}
+
+tensor_return_value_t DropoutDeviceOperation::create_output_tensors(
+    const operation_attributes_t& args, const tensor_args_t& tensor_args) {
+    if (tensor_args.preallocated_output.has_value()) {
+        return *tensor_args.preallocated_output;
+    }
+    return create_device_tensor(compute_output_specs(args, tensor_args), tensor_args.input.device());
+}
+
+tt::stl::hash::hash_t DropoutDeviceOperation::compute_program_hash(
+    const operation_attributes_t& args, const tensor_args_t& tensor_args) {
+    const auto& input_tensor = tensor_args.input;
+    const auto& input_shape = input_tensor.legacy_shape();
+    auto args_without_seed = args;
+    args_without_seed.seed = 0;
+    auto program_factory = select_program_factory(args, tensor_args);
+    operation::Hash hash = operation::hash_operation<DropoutDeviceOperation>(
+        args_without_seed,
+        program_factory.index(),
+        input_tensor.dtype(),
+        std::get<DeviceStorage>(input_tensor.storage()).memory_config(),
+        compute_volume(input_shape));
```

Comment:
are you sure the volume matters and is not a part of the runtime args?

Reply:
I think I took it from the unary ops. Not sure 100%

Reply:
weird if shape matters in the hash here or I don't understand something

Reply:
Ill leave it as is for now. Because I got it from unary infra.